psi4/driver/mdi_engine.py

```diff
@@ -0,0 +1,417 @@
+
+
+class MDIEngine():
+    def __init__(self, scf_method, molecule=None):
+        """ Initialize an MDIEngine object for communication with MDI
+
+        Arguments:
+           scf_method: Method used when calculating energies or gradients
+        """
+
+        # Method used when the SCF command is received
+        self.scf_method = scf_method
+
+        # Molecule all MDI operations are performed on
+        if molecule is None:
+            self.molecule = psi4.core.get_active_molecule()
+        else:
+            self.molecule = molecule
+
+        # Most recent SCF energy
+        self.energy = 0.0
+
+        # Variables used when MDI sets a lattice of point charges
+        self.nlattice = 0  # number of lattice point charges
+        self.clattice = []  # list of lattice coordinates
+        self.lattice = []  # list of lattice charges
+        self.lattice_field = psi4.QMMM()  # Psi4 chargefield
+
+        # MPI variables
+        self.mpi_world = None
+        self.world_rank = 0
+
+        # Get correct intra-code MPI communicator
+        if use_mpi4py:
+            self.mpi_world = MDI_Get_Intra_Code_MPI_Comm()
+            self.world_rank = self.mpi_world.Get_rank()
+
+        # Accept a communicator to the driver code
+        self.comm = MDI_Accept_Communicator()
+
+        # Ensure that the molecule is using c1 symmetry
+        self.molecule.reset_point_group('c1')
+        self.molecule.fix_orientation(True)
+        self.molecule.fix_com(True)
+        self.molecule.reinterpret_coordentry(False)
+        self.molecule.update_geometry()
+
+        # Flag to stop listening for MDI commands
+        self.stop_listening = False
+
+        # Dictionary of all supported MDI commands
+        self.commands = {
+            "<NATOMS": self.send_natoms,
+            "<COORDS": self.send_coords,
+            "<CHARGES": self.send_charges,
+            "<ELEMENTS": self.send_elements,
+            "<MASSES": self.send_masses,
+            "<ENERGY": self.send_energy,
+            "<FORCES": self.send_forces,
+            ">COORDS": self.recv_coords,
+            ">NLATTICE": self.recv_nlattice,
+            ">CLATTICE": self.recv_clattice,
+            ">LATTICE": self.recv_lattice,
+            ">MASSES": self.recv_masses,
+            "SCF": self.run_scf,
+            "<DIMENSIONS": self.send_dimensions,
+            "<NCOMMANDS": self.send_ncommands,
+            "<COMMANDS": self.send_commands,
+            "<TOTCHARGE": self.send_total_charge,
+            ">TOTCHARGE": self.recv_total_charge,
+            "<ELEC_MULT": self.send_multiplicity,
+            ">ELEC_MULT": self.recv_multiplicity,
+            "EXIT": self.exit
+        }
+
+    def length_conversion(self):
+        """ Obtain the conversion factor between the geometry specification units and bohr
+        """
+        unit_name = self.molecule.units()
+        if unit_name == "Angstrom":
+            unit_conv = MDI_Conversion_Factor("bohr", "angstrom")
```

Comment:
for Angstrom, do you want the value by which psi is converting the stored coords (Ang) to run the method (Bohr)? that's from qcel and is stored in the psi4.core.Mol class. or do you really want the MDI conversion factor?